File: android-java/app/src/main/java/com/oursky/skeleton/helper/ViewTransition.java
package com.oursky.skeleton.helper;

import android.animation.Animator;
import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.support.annotation.Nullable;
import android.view.View;

@SuppressWarnings("WeakerAccess")
public class ViewTransition {
    public interface OnCompleteListener {
        void onComplete();
    }
    public enum Direction { UP, DOWN, LEFT, RIGHT }

    //region fade
    //---------------------------------------------------------------
    public static void fadeIn(final View v, int duration, final @Nullable OnCompleteListener cb) {
        v.setAlpha(0);
        v.setVisibility(View.VISIBLE);
        ObjectAnimator ani = ObjectAnimator.ofFloat(v, "alpha", 1.0f);
        ani.setDuration(duration);
        if (cb!=null) {
            ani.addListener(new Animator.AnimatorListener() {
                @Override
                public void onAnimationStart(Animator animator) {
                }
                @Override
                public void onAnimationEnd(Animator animator) {
                    cb.onComplete();
                }
                @Override
                public void onAnimationCancel(Animator animator) {
                }
                @Override
                public void onAnimationRepeat(Animator animator) {
                }
            });
        }
        ani.start();
    }
    public static void fadeOut(final View v, int duration, final @Nullable OnCompleteListener cb) {
        ObjectAnimator ani = ObjectAnimator.ofFloat(v, "alpha", 0.0f);
        ani.setDuration(duration);
        ani.addListener(new Animator.AnimatorListener() {
            @Override
            public void onAnimationStart(Animator animator) {
            }
            @Override
            public void onAnimationEnd(Animator animator) {
                v.setVisibility(View.INVISIBLE);
                v.setAlpha(1.0f);
                if (cb!=null) cb.onComplete();
            }
            @Override
            public void onAnimationCancel(Animator animator) {
            }
            @Override
            public void onAnimationRepeat(Animator animator) {
            }
        });
        ani.start();
    }
    public static void fade(View from, View to, int duration, final @Nullable OnCompleteListener cb) {
        fadeOut(from, duration, null);
        fadeIn(to, duration, cb);
    }
    //---------------------------------------------------------------
    //endregion

    //region slide
    //---------------------------------------------------------------
    public static void slideIn(final View v, Direction dir, int translation, int duration, final @Nullable OnCompleteListener cb) {
        Animator.AnimatorListener animatorListener = new Animator.AnimatorListener() {
            @Override
            public void onAnimationStart(Animator animator) {
            }
            @Override
            public void onAnimationEnd(Animator animator) {
                if (cb!=null) cb.onComplete();
            }
            @Override
            public void onAnimationCancel(Animator animator) {
            }
            @Override
            public void onAnimationRepeat(Animator animator) {
            }
        };
        switch (dir) {
            case UP: {
                v.setTranslationX(0);
                v.setTranslationY(translation);
                v.setVisibility(View.VISIBLE);
                ObjectAnimator trans = ObjectAnimator.ofFloat(v, "translationY", 0.0f);
                trans.setDuration(duration);
                if (cb!=null) {
                    trans.addListener(animatorListener);
                }
                trans.start();
                break;
            }
            case DOWN: {
                v.setTranslationX(0);
                v.setTranslationY(-translation);
                v.setVisibility(View.VISIBLE);
                ObjectAnimator trans = ObjectAnimator.ofFloat(v, "translationY", 0.0f);
                trans.setDuration(duration);
                if (cb!=null) {
                    trans.addListener(animatorListener);
                }
                trans.start();
                break;
            }
            case LEFT: {
                v.setTranslationX(translation);
                v.setTranslationY(0);
                v.setVisibility(View.VISIBLE);
                ObjectAnimator trans = ObjectAnimator.ofFloat(v, "translationX", 0.0f);
                trans.setDuration(duration);
                if (cb!=null) {
                    trans.addListener(animatorListener);
                }
                trans.start();
                break;
            }
            case RIGHT: {
                v.setTranslationX(-translation);
                v.setTranslationY(0);
                v.setVisibility(View.VISIBLE);
                ObjectAnimator trans = ObjectAnimator.ofFloat(v, "translationX", 0.0f);
                trans.setDuration(duration);
                if (cb!=null) {
                    trans.addListener(animatorListener);
                }
                trans.start();
                break;
            }
        }
    }
    public static void slideOut(final View v, Direction dir, int translation, int duration, final @Nullable OnCompleteListener cb) {
        Animator.AnimatorListener animatorListener = new Animator.AnimatorListener() {
            @Override
            public void onAnimationStart(Animator animator) {
            }
            @Override
            public void onAnimationEnd(Animator animator) {
                v.setVisibility(View.INVISIBLE);
                v.setTranslationX(0);
                v.setTranslationY(0);
                if(cb!=null) cb.onComplete();
            }
            @Override
            public void onAnimationCancel(Animator animator) {
            }
            @Override
            public void onAnimationRepeat(Animator animator) {
            }
        };
        v.setTranslationX(0);
        v.setTranslationY(0);
        switch (dir) {
            case UP: {
                ObjectAnimator trans = ObjectAnimator.ofFloat(v, "translationY", -translation);
                trans.setDuration(duration);
                trans.addListener(animatorListener);
                trans.start();
                break;
            }
            case DOWN: {
                ObjectAnimator trans = ObjectAnimator.ofFloat(v, "translationY", translation);
                trans.setDuration(duration);
                trans.addListener(animatorListener);
                trans.start();
                break;
            }
            case LEFT: {
                ObjectAnimator trans = ObjectAnimator.ofFloat(v, "translationX", -translation);
                trans.setDuration(duration);
                trans.addListener(animatorListener);
                trans.start();
                break;
            }
            case RIGHT: {
                ObjectAnimator trans = ObjectAnimator.ofFloat(v, "translationX", translation);
                trans.setDuration(duration);
                trans.addListener(animatorListener);
                trans.start();
                break;
            }
        }
    }
    public static void slide(View from, View to, Direction dir, int translation, int duration, final @Nullable OnCompleteListener cb) {
        slideOut(from, dir, translation, duration, null);
        slideIn(to, dir, translation, duration, cb);
    }
    //---------------------------------------------------------------
    //endregion

    //region fadeslide
    //---------------------------------------------------------------
    public static void fadeslideIn(final View v, Direction dir, int translation, int duration, final @Nullable OnCompleteListener cb) {
        Animator.AnimatorListener animatorListener = new Animator.AnimatorListener() {
            @Override
            public void onAnimationStart(Animator animator) {
            }
            @Override
            public void onAnimationEnd(Animator animator) {
                if (cb!=null) cb.onComplete();
            }
            @Override
            public void onAnimationCancel(Animator animator) {
            }
            @Override
            public void onAnimationRepeat(Animator animator) {
            }
        };
        switch (dir) {
            case UP: {
                v.setTranslationX(0);
                v.setTranslationY(translation);
                v.setAlpha(0);
                v.setVisibility(View.VISIBLE);
                ObjectAnimator trans = ObjectAnimator.ofFloat(v, "translationY", 0.0f);
                ObjectAnimator alpha = ObjectAnimator.ofFloat(v, "alpha", 1.0f);
                trans.setDuration(duration);
                alpha.setDuration(duration);
                AnimatorSet aniSet = new AnimatorSet();
                if (cb!=null) {
                    aniSet.addListener(animatorListener);
                }
                aniSet.play(trans).with(alpha);
                aniSet.start();
                break;
            }
            case DOWN: {
                v.setTranslationX(0);
                v.setTranslationY(-translation);
                v.setAlpha(0);
                v.setVisibility(View.VISIBLE);
                ObjectAnimator trans = ObjectAnimator.ofFloat(v, "translationY", 0.0f);
                ObjectAnimator alpha = ObjectAnimator.ofFloat(v, "alpha", 1.0f);
                trans.setDuration(duration);
                alpha.setDuration(duration);
                AnimatorSet aniSet = new AnimatorSet();
                if (cb!=null) {
                    aniSet.addListener(animatorListener);
                }
                aniSet.play(trans).with(alpha);
                aniSet.start();
                break;
            }
            case LEFT: {
                v.setTranslationX(translation);
                v.setTranslationY(0);
                v.setAlpha(0);
                v.setVisibility(View.VISIBLE);
                ObjectAnimator trans = ObjectAnimator.ofFloat(v, "translationX", 0.0f);
                ObjectAnimator alpha = ObjectAnimator.ofFloat(v, "alpha", 1.0f);
                trans.setDuration(duration);
                alpha.setDuration(duration);
                AnimatorSet aniSet = new AnimatorSet();
                if (cb!=null) {
                    aniSet.addListener(animatorListener);
                }
                aniSet.play(trans).with(alpha);
                aniSet.start();
                break;
            }
            case RIGHT: {
                v.setTranslationX(-translation);
                v.setTranslationY(0);
                v.setAlpha(0);
                v.setVisibility(View.VISIBLE);
                ObjectAnimator trans = ObjectAnimator.ofFloat(v, "translationX", 0.0f);
                ObjectAnimator alpha = ObjectAnimator.ofFloat(v, "alpha", 1.0f);
                trans.setDuration(duration);
                alpha.setDuration(duration);
                AnimatorSet aniSet = new AnimatorSet();
                if (cb!=null) {
                    aniSet.addListener(animatorListener);
                }
                aniSet.play(trans).with(alpha);
                aniSet.start();
                break;
            }
        }
    }
    public static void fadeslideOut(final View v, Direction dir, int translation, int duration, final @Nullable OnCompleteListener cb) {
        Animator.AnimatorListener animatorListener = new Animator.AnimatorListener() {
            @Override
            public void onAnimationStart(Animator animator) {
            }
            @Override
            public void onAnimationEnd(Animator animator) {
                v.setVisibility(View.INVISIBLE);
                v.setTranslationX(0);
                v.setTranslationY(0);
                v.setAlpha(1.0f);
                if(cb!=null) cb.onComplete();
            }
            @Override
            public void onAnimationCancel(Animator animator) {
            }
            @Override
            public void onAnimationRepeat(Animator animator) {
            }
        };
        v.setTranslationX(0);
        v.setTranslationY(0);
        v.setAlpha(1.0f);
        switch (dir) {
            case UP: {
                ObjectAnimator trans = ObjectAnimator.ofFloat(v, "translationY", -translation);
                ObjectAnimator alpha = ObjectAnimator.ofFloat(v, "alpha", 0.0f);
                trans.setDuration(duration);
                alpha.setDuration(duration);
                AnimatorSet aniSet = new AnimatorSet();
                aniSet.addListener(animatorListener);
                aniSet.play(trans).with(alpha);
                aniSet.start();
                break;
            }
            case DOWN: {
                ObjectAnimator trans = ObjectAnimator.ofFloat(v, "translationY", translation);
                ObjectAnimator alpha = ObjectAnimator.ofFloat(v, "alpha", 0.0f);
                trans.setDuration(duration);
                alpha.setDuration(duration);
                AnimatorSet aniSet = new AnimatorSet();
                aniSet.addListener(animatorListener);
                aniSet.play(trans).with(alpha);
                aniSet.start();
                break;
            }
            case LEFT: {
                ObjectAnimator trans = ObjectAnimator.ofFloat(v, "translationX", -translation);
                ObjectAnimator alpha = ObjectAnimator.ofFloat(v, "alpha", 0.0f);
                trans.setDuration(duration);
                alpha.setDuration(duration);
                AnimatorSet aniSet = new AnimatorSet();
                aniSet.addListener(animatorListener);
                aniSet.play(trans).with(alpha);
                aniSet.start();
                break;
            }
            case RIGHT: {
                ObjectAnimator trans = ObjectAnimator.ofFloat(v, "translationX", translation);
                ObjectAnimator alpha = ObjectAnimator.ofFloat(v, "alpha", 0.0f);
                trans.setDuration(duration);
                alpha.setDuration(duration);
                AnimatorSet aniSet = new AnimatorSet();
                aniSet.addListener(animatorListener);
                aniSet.play(trans).with(alpha);
                aniSet.start();
                break;
            }
        }
    }
    public static void fadeslide(View from, View to, Direction dir, int translation, int duration, final OnCompleteListener cb) {
        fadeslideOut(from, dir, translation, duration, null);
        fadeslideIn(to, dir, translation, duration, cb);
    }
    //---------------------------------------------------------------
    //endregion
}

File: android-kotlin/app/src/main/java/com/oursky/skeleton/helper/ViewTransition.kt
package com.oursky.skeleton.helper

import android.animation.Animator
import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.view.View

@Suppress("unused", "MemberVisibilityCanBePrivate")
object ViewTransition {
    enum class Direction {
        UP, DOWN, LEFT, RIGHT
    }

    //region fade
    //---------------------------------------------------------------
    fun fadeIn(v: View, duration: Int, cb: (() -> Unit)? = null) {
        v.alpha = 0f
        v.visibility = View.VISIBLE
        val ani = ObjectAnimator.ofFloat(v, "alpha", 1.0f)
        ani.duration = duration.toLong()
        if (cb != null) {
            ani.addListener(object : Animator.AnimatorListener {
                override fun onAnimationStart(animator: Animator) {}
                override fun onAnimationEnd(animator: Animator) {
                    cb.invoke()
                }
                override fun onAnimationCancel(animator: Animator) {}
                override fun onAnimationRepeat(animator: Animator) {}
            })
        }
        ani.start()
    }
    fun fadeOut(v: View, duration: Int, cb: (() -> Unit)? = null) {
        val ani = ObjectAnimator.ofFloat(v, "alpha", 0.0f)
        ani.duration = duration.toLong()
        ani.addListener(object : Animator.AnimatorListener {
            override fun onAnimationStart(animator: Animator) {}
            override fun onAnimationEnd(animator: Animator) {
                v.visibility = View.INVISIBLE
                v.alpha = 1.0f
                cb?.invoke()
            }
            override fun onAnimationCancel(animator: Animator) {}
            override fun onAnimationRepeat(animator: Animator) {}
        })
        ani.start()
    }
    fun fade(from: View, to: View, duration: Int, cb: (() -> Unit)? = null) {
        fadeOut(from, duration, null)
        fadeIn(to, duration, cb)
    }
    //---------------------------------------------------------------
    //endregion

    //region slide
    //---------------------------------------------------------------
    fun slideIn(v: View, dir: Direction, translation: Int, duration: Int, cb: (() -> Unit)? = null) {
        val animatorListener = object : Animator.AnimatorListener {
            override fun onAnimationStart(animator: Animator) {}
            override fun onAnimationEnd(animator: Animator) {
                cb?.invoke()
            }
            override fun onAnimationCancel(animator: Animator) {}
            override fun onAnimationRepeat(animator: Animator) {}
        }
        when (dir) {
            ViewTransition.Direction.UP -> {
                v.translationX = 0f
                v.translationY = translation.toFloat()
                v.visibility = View.VISIBLE
                val trans = ObjectAnimator.ofFloat(v, "translationY", 0.0f)
                trans.duration = duration.toLong()
                if (cb != null) {
                    trans.addListener(animatorListener)
                }
                trans.start()
            }
            ViewTransition.Direction.DOWN -> {
                v.translationX = 0f
                v.translationY = (-translation).toFloat()
                v.visibility = View.VISIBLE
                val trans = ObjectAnimator.ofFloat(v, "translationY", 0.0f)
                trans.duration = duration.toLong()
                if (cb != null) {
                    trans.addListener(animatorListener)
                }
                trans.start()
            }
            ViewTransition.Direction.LEFT -> {
                v.translationX = translation.toFloat()
                v.translationY = 0f
                v.visibility = View.VISIBLE
                val trans = ObjectAnimator.ofFloat(v, "translationX", 0.0f)
                trans.duration = duration.toLong()
                if (cb != null) {
                    trans.addListener(animatorListener)
                }
                trans.start()
            }
            ViewTransition.Direction.RIGHT -> {
                v.translationX = (-translation).toFloat()
                v.translationY = 0f
                v.visibility = View.VISIBLE
                val trans = ObjectAnimator.ofFloat(v, "translationX", 0.0f)
                trans.duration = duration.toLong()
                if (cb != null) {
                    trans.addListener(animatorListener)
                }
                trans.start()
            }
        }
    }
    fun slideOut(v: View, dir: Direction, translation: Int, duration: Int, cb: (() -> Unit)? = null) {
        val animatorListener = object : Animator.AnimatorListener {
            override fun onAnimationStart(animator: Animator) {}
            override fun onAnimationEnd(animator: Animator) {
                v.visibility = View.INVISIBLE
                v.translationX = 0f
                v.translationY = 0f
                if (cb != null) cb()
            }

            override fun onAnimationCancel(animator: Animator) {}
            override fun onAnimationRepeat(animator: Animator) {}
        }
        v.translationX = 0f
        v.translationY = 0f
        when (dir) {
            ViewTransition.Direction.UP -> {
                val trans = ObjectAnimator.ofFloat(v, "translationY", -translation.toFloat())
                trans.duration = duration.toLong()
                trans.addListener(animatorListener)
                trans.start()
            }
            ViewTransition.Direction.DOWN -> {
                val trans = ObjectAnimator.ofFloat(v, "translationY", translation.toFloat())
                trans.duration = duration.toLong()
                trans.addListener(animatorListener)
                trans.start()
            }
            ViewTransition.Direction.LEFT -> {
                val trans = ObjectAnimator.ofFloat(v, "translationX", -translation.toFloat())
                trans.duration = duration.toLong()
                trans.addListener(animatorListener)
                trans.start()
            }
            ViewTransition.Direction.RIGHT -> {
                val trans = ObjectAnimator.ofFloat(v, "translationX", translation.toFloat())
                trans.duration = duration.toLong()
                trans.addListener(animatorListener)
                trans.start()
            }
        }
    }
    fun slide(from: View, to: View, dir: Direction, translation: Int, duration: Int, cb: (() -> Unit)?) {
        slideOut(from, dir, translation, duration, null)
        slideIn(to, dir, translation, duration, cb)
    }
    //---------------------------------------------------------------
    //endregion

    //region fadeslide
    //---------------------------------------------------------------
    fun fadeslideIn(v: View, dir: Direction, translation: Int, duration: Int, cb: (() -> Unit)? = null) {
        val animatorListener = object : Animator.AnimatorListener {
            override fun onAnimationStart(animator: Animator) {}
            override fun onAnimationEnd(animator: Animator) {
                cb?.invoke()
            }
            override fun onAnimationCancel(animator: Animator) {}
            override fun onAnimationRepeat(animator: Animator) {}
        }
        when (dir) {
            ViewTransition.Direction.UP -> {
                v.translationX = 0f
                v.translationY = translation.toFloat()
                v.alpha = 0f
                v.visibility = View.VISIBLE
                val trans = ObjectAnimator.ofFloat(v, "translationY", 0.0f)
                val alpha = ObjectAnimator.ofFloat(v, "alpha", 1.0f)
                trans.duration = duration.toLong()
                alpha.duration = duration.toLong()
                val aniSet = AnimatorSet()
                if (cb != null) {
                    aniSet.addListener(animatorListener)
                }
                aniSet.play(trans).with(alpha)
                aniSet.start()
            }
            ViewTransition.Direction.DOWN -> {
                v.translationX = 0f
                v.translationY = (-translation).toFloat()
                v.alpha = 0f
                v.visibility = View.VISIBLE
                val trans = ObjectAnimator.ofFloat(v, "translationY", 0.0f)
                val alpha = ObjectAnimator.ofFloat(v, "alpha", 1.0f)
                trans.duration = duration.toLong()
                alpha.duration = duration.toLong()
                val aniSet = AnimatorSet()
                if (cb != null) {
                    aniSet.addListener(animatorListener)
                }
                aniSet.play(trans).with(alpha)
                aniSet.start()
            }
            ViewTransition.Direction.LEFT -> {
                v.translationX = translation.toFloat()
                v.translationY = 0f
                v.alpha = 0f
                v.visibility = View.VISIBLE
                val trans = ObjectAnimator.ofFloat(v, "translationX", 0.0f)
                val alpha = ObjectAnimator.ofFloat(v, "alpha", 1.0f)
                trans.duration = duration.toLong()
                alpha.duration = duration.toLong()
                val aniSet = AnimatorSet()
                if (cb != null) {
                    aniSet.addListener(animatorListener)
                }
                aniSet.play(trans).with(alpha)
                aniSet.start()
            }
            ViewTransition.Direction.RIGHT -> {
                v.translationX = (-translation).toFloat()
                v.translationY = 0f
                v.alpha = 0f
                v.visibility = View.VISIBLE
                val trans = ObjectAnimator.ofFloat(v, "translationX", 0.0f)
                val alpha = ObjectAnimator.ofFloat(v, "alpha", 1.0f)
                trans.duration = duration.toLong()
                alpha.duration = duration.toLong()
                val aniSet = AnimatorSet()
                if (cb != null) {
                    aniSet.addListener(animatorListener)
                }
                aniSet.play(trans).with(alpha)
                aniSet.start()
            }
        }
    }
    fun fadeslideOut(v: View, dir: Direction, translation: Int, duration: Int, cb: (() -> Unit)? = null) {
        val animatorListener = object : Animator.AnimatorListener {
            override fun onAnimationStart(animator: Animator) {}
            override fun onAnimationEnd(animator: Animator) {
                v.visibility = View.INVISIBLE
                v.translationX = 0f
                v.translationY = 0f
                v.alpha = 1.0f
                cb?.invoke()
            }
            override fun onAnimationCancel(animator: Animator) {}
            override fun onAnimationRepeat(animator: Animator) {}
        }
        v.translationX = 0f
        v.translationY = 0f
        v.alpha = 1.0f
        when (dir) {
            ViewTransition.Direction.UP -> {
                val trans = ObjectAnimator.ofFloat(v, "translationY", -translation.toFloat())
                val alpha = ObjectAnimator.ofFloat(v, "alpha", 0.0f)
                trans.duration = duration.toLong()
                alpha.duration = duration.toLong()
                val aniSet = AnimatorSet()
                aniSet.addListener(animatorListener)
                aniSet.play(trans).with(alpha)
                aniSet.start()
            }
            ViewTransition.Direction.DOWN -> {
                val trans = ObjectAnimator.ofFloat(v, "translationY", translation.toFloat())
                val alpha = ObjectAnimator.ofFloat(v, "alpha", 0.0f)
                trans.duration = duration.toLong()
                alpha.duration = duration.toLong()
                val aniSet = AnimatorSet()
                aniSet.addListener(animatorListener)
                aniSet.play(trans).with(alpha)
                aniSet.start()
            }
            ViewTransition.Direction.LEFT -> {
                val trans = ObjectAnimator.ofFloat(v, "translationX", -translation.toFloat())
                val alpha = ObjectAnimator.ofFloat(v, "alpha", 0.0f)
                trans.duration = duration.toLong()
                alpha.duration = duration.toLong()
                val aniSet = AnimatorSet()
                aniSet.addListener(animatorListener)
                aniSet.play(trans).with(alpha)
                aniSet.start()
            }
            ViewTransition.Direction.RIGHT -> {
                val trans = ObjectAnimator.ofFloat(v, "translationX", translation.toFloat())
                val alpha = ObjectAnimator.ofFloat(v, "alpha", 0.0f)
                trans.duration = duration.toLong()
                alpha.duration = duration.toLong()
                val aniSet = AnimatorSet()
                aniSet.addListener(animatorListener)
                aniSet.play(trans).with(alpha)
                aniSet.start()
            }
        }
    }
    fun fadeslide(from: View, to: View, dir: Direction, translation: Int, duration: Int, cb: (() -> Unit)? = null) {
        fadeslideOut(from, dir, translation, duration, null)
        fadeslideIn(to, dir, translation, duration, cb)
    }
    //---------------------------------------------------------------
    //endregion
}
